packages/expo-sqlite/android/src/main/java/expo/modules/sqlite/SQLiteModule.kt

```diff
@@ -0,0 +1,170 @@
+import java.util.*
+
+class SQLiteModule(private val mContext: Context) : ExportedModule(mContext) {
+  companion object {
+    private const val DEBUG_MODE = false
+    private val TAG = SQLiteModule::class.java.simpleName
+    private val EMPTY_ROWS = arrayOf<Array<Any?>>()
+    private val EMPTY_COLUMNS = arrayOf<String?>()
+    private val EMPTY_RESULT = SQLitePluginResult(EMPTY_ROWS, EMPTY_COLUMNS, 0, 0, null)
+    private val DATABASES: MutableMap<String, SQLiteDatabase?> = HashMap()
+  }
+
+  override fun getName(): String {
+    return "ExponentSQLite"
+  }
+
+  @ExpoMethod
+  fun exec(dbName: String, queries: ArrayList<ArrayList<Any>>, readOnly: Boolean, promise: Promise) {
+    try {
+      val db = getDatabase(dbName)
+      val results = queries.map { sqlQuery ->
+        val sql = sqlQuery[0] as String
+        val bindArgs = convertParamsToStringArray(sqlQuery[1] as ArrayList<Any?>)
+        try {
+          if (isSelect(sql)) {
+            doSelectInBackgroundAndPossiblyThrow(sql, bindArgs, db)
+          } else { // update/insert/delete
+            if (readOnly) {
+              SQLitePluginResult(EMPTY_ROWS, EMPTY_COLUMNS, 0, 0, ReadOnlyException())
+            } else {
+              doUpdateInBackgroundAndPossiblyThrow(sql, bindArgs, db)
+            }
+          }
+        } catch (e: Throwable) {
+          if (DEBUG_MODE) {
+            e.printStackTrace()
+          }
+          SQLitePluginResult(EMPTY_ROWS, EMPTY_COLUMNS, 0, 0, e)
+        }
+      }
+      val data = pluginResultsToPrimitiveData(results)
+      promise.resolve(data)
+    } catch (e: Exception) {
+      promise.reject("SQLiteError", e)
+    }
+  }
+
+  @ExpoMethod
+  fun close(dbName: String, promise: Promise) {
+    DATABASES
+        .remove(dbName)
+        ?.close()
+    promise.resolve(null)
+  }
+
+  // do a update/delete/insert operation
+  private fun doUpdateInBackgroundAndPossiblyThrow(
+      sql: String,
+      bindArgs: Array<String?>?,
+      db: SQLiteDatabase
+  ): SQLitePluginResult {
+    return db.compileStatement(sql).use { statement ->
+      if (bindArgs != null) {
+        for (i in bindArgs.size downTo 1) {
+          if (bindArgs[i - 1] == null) {
+            statement.bindNull(i)
+          } else {
+            statement.bindString(i, bindArgs[i - 1])
+          }
+        }
+      }
+      if (isInsert(sql)) {
+        val insertId = statement.executeInsert()
+        val rowsAffected = if (insertId >= 0) 1 else 0
+        SQLitePluginResult(EMPTY_ROWS, EMPTY_COLUMNS, rowsAffected, insertId, null)
+      } else if (isDelete(sql) || isUpdate(sql)) {
+        val rowsAffected = statement.executeUpdateDelete()
+        SQLitePluginResult(EMPTY_ROWS, EMPTY_COLUMNS, rowsAffected, 0, null)
+      } else {
+        // in this case, we don't need rowsAffected or insertId, so we can have a slight
+        // perf boost by just executing the query
+        statement.execute()
+        EMPTY_RESULT
+      }
+    }
```

Comment:
~Don't we have to call `statement.close()` at the end of the processing?~
There was something like this in JAVA:
```java
} finally {
  if (statement != null) {
    statement.close(
  }
}
```
I've found that [`SQLiteStatemenet`](https://developer.android.com/reference/android/database/sqlite/SQLiteStatement) implements [`AutoCLoseable`](https://developer.android.com/reference/java/lang/AutoCloseable), so that's not necessary indeed 👍

Reply:
"The close() method of an AutoCloseable object is called automatically when exiting a try-with-resources block for which the object has been declared in the resource specification header. This construction ensures prompt release, avoiding resource exhaustion exceptions and errors that may otherwise occur."

So, you need to ensure that the cursors will be closed. You can do it automatically if you used the try-with-resources construction in Java or the `use ` function from the Kotlin library.  But it doesn't mean, that closing the cursor is unnecessary.